### PR TITLE
Nimbus light client integration with status-go rebased version

### PR DESCRIPTION
**Pull request by vitvly**

### CODE REVIEW

**Comment by arnetheduck** [comment]
This version still looks like it's missing the separate library "main" file, suggested here: https://github.com/status-im/nimbus-eth1/pull/1572#issuecomment-1838885345

With that approach, we can also un-declare metrics for that file specifically and the rest of the nimbus-eth1 project is not affected.

File: nimbus_verified_proxy/nimbus_verified_proxy.nim [hunk 1]
```diff
@@ -39,14 +41,22 @@ func getConfiguredChainId(networkMetadata: Eth2NetworkMetadata): Quantity =
   else:
     return networkMetadata.cfg.DEPOSIT_CHAIN_ID.Quantity
 
-proc run(config: VerifiedProxyConf) {.raises: [CatchableError].} =
+proc run*(config: VerifiedProxyConf, callbacks: varargs[OnHeaderCallback]) {.raises: [CatchableError], gcsafe.} =
```
Comment (arnetheduck) [inline nimbus_verified_proxy/nimbus_verified_proxy.nim:44]:
```suggestion
proc run*(config: VerifiedProxyConf, headerCallback: OnHeaderCallback = nil) {.raises: [CatchableError], gcsafe.} =
```

File: nimbus_verified_proxy/nimbus_verified_proxy.nim [hunk 2]
```diff
@@ -167,6 +177,13 @@ proc run(config: VerifiedProxyConf) {.raises: [CatchableError].} =
       when lcDataFork > LightClientDataFork.None:
         info "New LC finalized header",
           finalized_header = shortLog(forkyHeader)
+        if headerCallback != nil:
+          notice "### Invoking finalizedHeaderCallback"
```
Comment (arnetheduck) [inline nimbus_verified_proxy/nimbus_verified_proxy.nim:181]: debug leftover?

File: nimbus_verified_proxy/nimbus_verified_proxy.nim [hunk 3]
```diff
@@ -253,10 +277,9 @@ proc run(config: VerifiedProxyConf) {.raises: [CatchableError].} =
   while true:
     poll()
 
-when isMainModule:
+when isMainModule and appType != "lib":
```
Comment (arnetheduck) [inline nimbus_verified_proxy/nimbus_verified_proxy.nim:280]: no longer needed

File: nimbus_verified_proxy/nimbus_verified_proxy.nim (hunk 1 above)
Comment (vitvly) [reply nimbus_verified_proxy/nimbus_verified_proxy.nim]: Replaced it with `ptr Context` param.

File: nimbus_verified_proxy/nimbus_verified_proxy.nim (hunk 2 above)
Comment (vitvly) [reply nimbus_verified_proxy/nimbus_verified_proxy.nim]: Removed.

File: nimbus_verified_proxy/nimbus_verified_proxy.nim (hunk 3 above)
Comment (vitvly) [reply nimbus_verified_proxy/nimbus_verified_proxy.nim]: Removed.

**Comment by vitvly** [comment]
Rebased with latest master, added cleanup code.

**Comment by vitvly** [comment]
Changed nim-confutils dependency to point to updated upstream.

**Review by kdeme — COMMENTED** [review]
Not so much a review, but I'm more curious about the planned architecture here.

nimbus_verified_proxy is supposed to be the trustless EL json-rpc API drop-in replacement. But I noticed in this PR also that new optimistic and finalized headers are being passed along via a callback. What is the architectural idea here exactly?

**Comment by vitvly** [comment]
For now the headers are only used to extract block roots in status-go, in order to persist sync progress. Otherwise not needed (yet?).

**Comment by kdeme** [comment]
> For now the headers are only used to extract block roots in status-go, in order to persist sync progress. Otherwise not needed (yet?).

Do you mean, so that when restarting, you can provided a much newer block root, to have a very short sync?

If so, this is a feature that I've been wanting to add to the nimbus_verified_proxy.

**Comment by vitvly** [comment]
> > For now the headers are only used to extract block roots in status-go, in order to persist sync progress. Otherwise not needed (yet?).
> 
> Do you mean, so that when restarting you can provide a much newer block root in order to have a very short sync?
> 
> If so, this is a feature that I've been wanting to add to the nimbus_verified_proxy.

Yes, it'll check finalized headers for a `state_root` value and store it in status-go DB. Would be great to have it in nimbus_verified_proxy i think.

**Pull request merged**